virt-controller, vm: Extract VM net logic out of the main flow

EdDev opened the pull request:
### What this PR does

The network logic that existed in the VM controller flow is extracted
out to be owned by the network SIG.

The integration point between the network logic and the VM controller
main flow has been generalized to fit other logical code groups to
follow the same path.

The following code components have been introduced in this change:
- The integration point at the VM controller flow is implemented through
  a pool of `synchronizers` which are executed in the controller flow by
  calling their `Sync` method.
  The objects themself are created at setup with the relevant data that
  they require (e.g. clientset, object stores, FG configuration).
- The concrete objects that implement the logic are a kind of
  sub-controllers, focused on specific logic (e.g. network) and
  completely decoupled and unaware of other logic.

The extraction and decoupling allowed to add unit test coverage to the
VM controller network logic.

<!-- (optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: -->
Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->

### Special notes for your reviewer

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

/sig code-quality

EdDev commented:
/cc @orelmisan @ormergi

xpivarc commented:
/cc @acardace 
You were looking at something similar, can you have a look and maybe sync with Ed?

EdDev commented:
change: Rebase

ormergi commented on pkg/virt-controller/watch/vm.go:
```diff
@@ -3135,31 +3141,17 @@ func (c *VMController) sync(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachin
-		if err := c.vmiInterfacesPatch(&vmiCopy.Spec, vmi); err != nil {
-			return vm, &syncErrorImpl{fmt.Errorf("Error encountered when trying to patch vmi: %v", err), FailedUpdateErrorReason}, nil
+	for _, syncer := range c.Synchronizers {
+		resVM, errSync := syncer.Sync(vmCopy, vmi)
```
Not sure what `resVM` means, please consider renaming it to something like syncedVM/updatedVM.

nit: I would rename `errSync` to plain `err`.

EdDev replied:
> Not sure what `resVM` means, please consider renaming it to something like syncedVM/updatedVM.

Done
 
> nit: I would rename `errSync` to plain `err`.

I prefer not to shadow the upper `err`.

ormergi commented on pkg/virt-controller/watch/vm.go:
```diff
@@ -3135,31 +3141,17 @@ func (c *VMController) sync(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachin
+			if errors.As(errSync, &errWithReason) {
+				return vm, errWithReason, nil
+			}
+			return vm, &syncErrorImpl{fmt.Errorf("unsupported error"), "UnsupportedSyncError"}, nil
```
There is the `UnsupportedSyncError` reason, but  it overrides the actual error message with "unsupported error", is it intentional?
If so, shouldn't it surface the actual error message?

EdDev replied:
Done

ormergi reviewed (commented):
Thanks for the PR awesome work!

orelmisan commented on pkg/virt-controller/network/vmcontroller_test.go:
```diff
@@ -0,0 +1,235 @@
+
+	It("has hotplug FG unset", func() {
+		c := network.NewVMNetController(fake.NewSimpleClientset(), stubClusterConfig{}, stubPodGetter{})
+		vmi := &v1.VirtualMachineInstance{}
```
nit: For consistency, please consider using `libvmi` as was done bellow.

orelmisan commented on pkg/virt-controller/network/vmcontroller_test.go:
```diff
@@ -0,0 +1,235 @@
+		c := network.NewVMNetController(fake.NewSimpleClientset(), stubClusterConfig{netHotplugEnabled: true}, podGetter)
+		Expect(c.Sync(vm, vmi)).To(Equal(vm))
+	},
+		Entry("there is no VMI", emptyVM, nil, stubPodGetter{}),
```
```suggestion
		Entry("The VM is not running - there is no VMI", emptyVM, nil, stubPodGetter{}),
```

orelmisan commented on pkg/virt-controller/network/vmcontroller_test.go:
```diff
@@ -0,0 +1,235 @@
+			stubPodGetter{},
+		),
+		Entry("the VM & VMI have no interfaces and no pod is found", emptyVM, libvmi.New(), stubPodGetter{}),
+	)
```
Is there value in testing the scenario when the VM is marked for deletion?

orelmisan commented on pkg/virt-controller/network/vmcontroller_test.go:
```diff
@@ -0,0 +1,235 @@
+		Expect(updatedVM.Spec.Template.Spec.Domain.Devices.Interfaces).To(Equal(updatedVMI.Spec.Domain.Devices.Interfaces))
+	})
+
+	It("sync succeeds to clear hotunplug interfaces", func() {
```
Could you please add a similar test case where the patch fails?

orelmisan commented on pkg/virt-controller/network/vmcontroller_test.go:
```diff
@@ -0,0 +1,235 @@
+		Expect(updatedVM).To(Equal(emptyVM))
+	})
+
+	It("sync succeeds to hotplug new interface", func() {
```
Could you please add a similar test case where the patch fails?

EdDev replied:
Please ack that this has been resolved so I can close the conversation.

orelmisan reviewed (commented):
Thank you for the PR @EdDev

EdDev commented on pkg/virt-controller/watch/vm.go:
```diff
@@ -249,6 +252,8 @@ type VMController struct {
 	statusUpdater          *status.VMStatusUpdater
 	clusterConfig          *virtconfig.ClusterConfig
 	hasSynced              func() bool
+
```
Nice catch.

Done.

EdDev replied:
It is used to wait for the informers to be "ready" before starting the controller main loop. I do not feel confident it can be removed.

orelmisan replied:
Fair enough.

EdDev commented on pkg/virt-controller/watch/vm.go:
```diff
@@ -3135,31 +3141,17 @@ func (c *VMController) sync(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachin
-
-		hasOrdinalIfaces, err := c.hasOrdinalNetworkInterfaces(vmi)
-		if err != nil {
-			return vm, &syncErrorImpl{fmt.Errorf("Error encountered when trying to check if VMI has interface with ordinal names (e.g.: eth1, eth2..): %v", err), HotPlugNetworkInterfaceErrorReason}, nil
```
Done

ormergi commented on pkg/virt-controller/network/vmcontroller.go:
```diff
@@ -0,0 +1,157 @@
+
+type podFromVMIGetter interface {
+	CurrentPod(vmi *v1.VirtualMachineInstance) (*k8sv1.Pod, error)
+}
```
The ` podFromVMIGetter` interface is used by `VMNetController`  thus I asked to bring it closer, sorry for not clarifying in the first place.

ormergi commented on pkg/virt-controller/network/vmcontroller_test.go:
```diff
@@ -0,0 +1,235 @@
+		_, err := clientset.KubevirtV1().VirtualMachineInstances(vmi.Namespace).Create(context.Background(), vmi, k8smetav1.CreateOptions{})
+		Expect(err).NotTo(HaveOccurred())
+
+		updatedVM, err := c.Sync(vm, vmi)
+		Expect(err).NotTo(HaveOccurred())
```
I guess this tests should be changed by the bug fix PR, to show the VM/VMI is not mutated, right?

EdDev replied:
I do not remember which test will fail when fixing the bug, but one does.

orelmisan commented on pkg/virt-controller/network/vmcontroller_test.go:
```diff
@@ -0,0 +1,328 @@
+
+	DescribeTable("sync does nothing when", func(vm *v1.VirtualMachine, vmi *v1.VirtualMachineInstance, podGetter stubPodGetter) {
+		c := network.NewVMNetController(fake.NewSimpleClientset(), stubClusterConfig{netHotplugEnabled: true}, podGetter)
+		originalVM := emptyVM.DeepCopy()
```
```suggestion
		originalVM := vm.DeepCopy()
```

EdDev replied:
Done

orelmisan commented on pkg/virt-controller/network/vmcontroller_test.go:
```diff
@@ -0,0 +1,328 @@
+			stubPodGetter{},
+		),
+		Entry("the VM & VMI have no interfaces and no pod is found", emptyVM, libvmi.New(), stubPodGetter{}),
+	)
```
Please consider adding a scenario where the VM and VMI specs are equal.

EdDev replied:
Done

orelmisan commented on pkg/virt-controller/network/vmcontroller_test.go:
```diff
@@ -0,0 +1,328 @@
+		Expect(updatedVM).To(Equal(originalVM))
+	})
+
+	It("sync fails when pod patch returns an error", func() {
```
```suggestion
	It("sync fails when VMI patch returns an error", func() {
```

EdDev replied:
Done

orelmisan replied:
It still says "pod".

EdDev replied:
Ahh, done.

orelmisan commented on pkg/virt-controller/network/vmcontroller_test.go:
```diff
@@ -0,0 +1,328 @@
+		originalVM := emptyVM.DeepCopy()
+		updatedVM, err := c.Sync(emptyVM, libvmi.New())
+		Expect(err).To(MatchError(isSyncErrorType, "syncError"))
+		Expect(updatedVM).To(Equal(originalVM))
```
Isn't a value returned from a function that also returned an error is undefined?
Same bellow.

EdDev replied:
This is how the current code behaves. It returns the current VM in case of failure.
I do not want to change this behavior as part of this PR.

Also, as a caller, I guess you can ignore the value. But in tests, we are suppose to express what actually happens. The fact it can contain information, I think it is important to express it.

orelmisan commented on pkg/virt-controller/network/vmcontroller_test.go:
```diff
@@ -0,0 +1,328 @@
+
+		vm = plugNetworkInterface(vm, "foonet")
+
+		// Simulate the existence of the VMI on the server (to allow the Sync to patch it).
+		_, err := clientset.KubevirtV1().VirtualMachineInstances(vmi.Namespace).Create(context.Background(), vmi, k8smetav1.CreateOptions{})
+		Expect(err).NotTo(HaveOccurred())
```
nit: The vmi could be passed to 
```go
clientset := fake.NewSimpleClientset(vmi)
```

and make this create operation redundant.
Same bellow.

EdDev replied:
Nice, but it will require to create the VMI before creating the net-controller which is odd.

orelmisan commented on pkg/virt-controller/network/vmcontroller_test.go:
```diff
@@ -0,0 +1,328 @@
+		pod := &k8sv1.Pod{
+			ObjectMeta: k8smetav1.ObjectMeta{Annotations: map[string]string{
+				networkv1.NetworkStatusAnnot: `[
+						{"interface":"eth0", "name":"red-net", "namespace": "default"},
+						{"interface":"net1", "name":"blue-net", "namespace": "default"}
```
nit: it could be nice to have the network names equal to what is defined on the VM/VMI.

EdDev replied:
Done

orelmisan replied:
The secondary network lacks the `-nad` suffix (since `name` represents the name of the NAD).

orelmisan reviewed (commented):
Thank you for the changes @EdDev 
Please see the inline comments.

EdDev commented on pkg/virt-controller/network/vmcontroller_test.go:
```diff
@@ -0,0 +1,328 @@
+	"context"
+	"errors"
+
+	k8sruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/testing"
```
Done

EdDev commented on pkg/virt-controller/network/vmcontroller_test.go:
```diff
@@ -0,0 +1,328 @@
+)
+
+var _ = Describe("VM Network Controller", func() {
+	emptyVM := &v1.VirtualMachine{Spec: v1.VirtualMachineSpec{Template: &v1.VirtualMachineInstanceTemplateSpec{}}}
```
Done

EdDev commented on pkg/virt-controller/network/vmcontroller_test.go:
```diff
@@ -0,0 +1,328 @@
+		updatedVMI, err := clientset.KubevirtV1().VirtualMachineInstances(vmi.Namespace).Get(context.Background(), vmi.Name, k8smetav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+
+		Expect(updatedVM.Spec.Template.Spec.Networks).To(Equal(updatedVMI.Spec.Networks))
+		Expect(updatedVM.Spec.Template.Spec.Domain.Devices.Interfaces).To(Equal(updatedVMI.Spec.Domain.Devices.Interfaces))
```
Both are outputs & side-effects of the sync.

IMO it does not matter, but no point in arguing about it.

Done

EdDev commented:
[change](https://github.com/kubevirt/kubevirt/compare/cb8a6ae3cd3ef9b7557b4dc6338473e3d07ab8f9..e7765e3f8e460a64c20b9745c0496b34577a348b): Answered review comments

EdDev commented on pkg/virt-controller/network/vmcontroller_test.go:
```diff
@@ -0,0 +1,235 @@
+	It("has hotplug FG unset", func() {
+		c := network.NewVMNetController(fake.NewSimpleClientset(), stubClusterConfig{}, stubPodGetter{})
+		vmi := &v1.VirtualMachineInstance{}
+		Expect(c.Sync(emptyVM, vmi)).To(Equal(emptyVM))
```
> 3. Please consider asserting that no action was performed against the API server:

That implies behavioral change, which has strong coupling with the implementation.
It is very similar to the mock usage which I am not friendly with.

orelmisan replied:
For the tests that have "sync does nothing" in their title, I think we should also assert that the VMI is unchanged (in case there is a VMI).
WDYT?

orelmisan replied:
The VMI is patched as part of hotplug/hotuplug operation.
Please consider adding an assertion that the VMI was not patched (where applicable).

EdDev replied:
A patch operation will fail if such an attempt is made, so what you are looking to assert is already implicitly done.

EdDev commented:
[change](https://github.com/kubevirt/kubevirt/compare/e7765e3f8e460a64c20b9745c0496b34577a348b..756659f3111788862b3dc0d1d2ab45b5ce976d79): Answered review.

orelmisan reviewed (approved):
Thank you @EdDev 
Looks great!

xpivarc commented on pkg/virt-controller/watch/vm.go:
```diff
@@ -231,6 +229,10 @@ func (p *authProxy) GetDataSource(namespace, name string) (*cdiv1.DataSource, er
 	return ds, nil
 }
 
+type synchronizer interface {
+	Sync(*virtv1.VirtualMachine, *virtv1.VirtualMachineInstance) (*virtv1.VirtualMachine, error)
+}
+
```
Should a synchronize always make a copy, does it returns copy... can you specify this

EdDev replied:
I initially thought about laying out some rules for the usage.
However, I realized this may cause some debate and we need to reach a wider agreement to have such rules. As I see it, this is flexible until we have at least 2 usages.

At its base, the only rules that really matters IMO are these:
- Avoid mutating the input objects (VM and VMI) and return the changes through the returned VM object.
- Avoid changing the `Status` of the returned VM object as it will not be consumed by the main flow.

The enforcement should be by the main flow. Even if the rules are not enforced by the concrete controller, the mainflow will enforce it.

If you still want me to add such rules as formal documentation to the interface, I can. My recommendation is to wait with it to a separate PR which will add another user for the interface.
I'm good either way.

xpivarc replied:
I would probably prefer if we don't try to solve this now and we just make an endpoint for the network code for now. The topic is a bit more complex than what is proposed imho. The `synchronizer` first should be called to validate hotplug can be performed, then we would set the condition or continue to call each `synchronizer` based on the validation.

EdDev replied:
I think you are attempting to inticipate how this will look like if and when you will take care of the other hotplug points. As this can go in many directions, I think we better first start with one, then see how it evolves when we add the others one by one.

The advantage to keep this generic is that we encourage other owners of existing codebases and future additions to think in this manner and not add more to the one-big-method.

EdDev replied:
> > I think you are attempting to inticipate how this will look
> 
> I actually argue that the code as of now is trying to forecast how this should look.... 

Yes, it tries to softly forecast how this should look given the existing user of it (the network sync). But this says nothing about the ability to change it per the need of a new user that will come in.

> Therefore I simply suggest to keep this simple and have just one function call instead of the current code block, no interface needed as of now.

The interface is technically needed in order to allow the decoupling between the controller flow and the network implementation.
The idea is to break the dependency, which is the purpose of interfaces.

> > The advantage to keep this generic is that we encourage other owners of existing codebases and future additions to think in this manner and not add more to the one-big-method.
> 
> Simple function has the same effect imho

Functions provide only abstraction layers and do not take care of code architecture concerns like decoupling and focused structured flows.
The sig-network is looking to have this decoupling in place to serve its own code need and encourage others to use the same when introducing new logic.

Do you see now a problem in the current generic suggestion that may break something? Do you think it will make it hard for the owners of this control flow to maintain this part?
I tried to make it as simple as possible so it will not be chalenging to maintain. In fact, it takes out the network logic so it should be easier for this codebase owner to maintain it now.

xpivarc replied:
Ok, so would we agree to keep the interface but remove the slice? In its current form it does suggest that it is ok just to add another synchronizer which we already know is not good enough. WDYT?

EdDev replied:
I think it is a reasonable compromise.
I will change it and as a follow up try to add another one in order to return the slice.

EdDev replied:
Done

xpivarc commented on pkg/virt-controller/watch/vm.go:
```diff
@@ -3141,31 +3145,17 @@ func (c *VMController) sync(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachin
+			}
+			return vm, &syncErrorImpl{fmt.Errorf("unsupported error: %v", errSync), "UnsupportedSyncError"}, nil
 		}
+		vmCopy.ObjectMeta = syncedVM.ObjectMeta
+		vmCopy.Spec = syncedVM.Spec
```
Why don't we modify the `vmCopy` directly?

EdDev replied:
I think the main flow (here) should enforce what can be changed and what cannot. It should not be possible for a concrete implementation to do whatever it wants.

Only the `ObjectData` and `Spec` can be modified on a `Sync` at this stage, and this is expressed here.

In general, I always prefer to have predictable code where input is not mutated and if there is a need to mutate something, I just see it on the return value.

xpivarc commented on pkg/virt-controller/watch/vm_test.go:
```diff
@@ -5517,6 +5519,53 @@ var _ = Describe("VirtualMachine", func() {
+					"Message": Equal(expectCond.Message),
+				}),
+			))
+			Expect(vm.Status.Conditions).To(HaveLen(expectCondCount))
```
Doesn't the count make this test fragile as well? Do we care how many condition are here?

EdDev replied:
It is a balance between fragile and expressing reality.
At the moment, this expresses that there are cases where there are more conditions. For this specific test we do not care about the details of those conditions, but we acknowledge there are. If we perform a change in the codebase and this one fails, it means we may have changes something unintentionally and we should make it is ok.

xpivarc replied:
Still I think what you care about here is only the one condition. It can well be that we remove/add or do both and I can't say why you would care here.

xpivarc commented on pkg/controller/controller.go:
```diff
@@ -114,6 +114,18 @@ const (
 	MigrationBackoffReason = "MigrationBackoff"
 )
 
+type PodCacheStore struct {
+	indexer cache.Indexer
+}
+
+func NewPodCacheStore(indexer cache.Indexer) *PodCacheStore {
+	return &PodCacheStore{indexer: indexer}
+}
+
+func (p *PodCacheStore) CurrentPod(vmi *v1.VirtualMachineInstance) (*k8sv1.Pod, error) {
+	return CurrentVMIPod(vmi, p.indexer)
+}
```
What value does this bring?

EdDev replied:
It allows the code that uses it to be testable. See the use of a stub to replace it in the tests below (`stubPodGetter`).

xpivarc replied:
Imho you could just keep the informer and use the `CurrentVMIPod` directly. Creating the `Store` for the tests as you need. This seems like additional complexity/boilerplate.

EdDev replied:
The `CurrentVMIPod` has logic that I do not want to depend on.
I do not want the network tests to know what an informer is, how its data is stored and so on.

For tests to be maintainable, they need to be simple and depend as little as possible on other logic.

The logic of `CurrentVMIPod` should be unit tested directly and there I guess an informer sore will be a good fit.

xpivarc replied:
Sorry meant to say "You could just keep the `Store` from informer...."

>For tests to be maintainable, they need to be simple and depend as little as possible on other logic.

In the case I described you would just pass relevant Store to the test so it would be simple, imho

>The logic of CurrentVMIPod should be unit tested directly and there I guess an informer store will be a good fit.

100% and funny enough it is not :)

EdDev replied:
> In the case I described you would just pass relevant Store to the test so it would be simple, imho

I do not thin it is simple. But even if it is, it assumes I know the logic if that method and I do not want to know how it works at all. I actually think it is wrong to depend on knowing how it works.

xpivarc commented on pkg/virt-controller/network/vmcontroller.go:
```diff
@@ -0,0 +1,160 @@
+}
+
+type VMNetController struct {
+	clientset     kubevirt.Interface
```
You may want to require only the `vmi` client to follow the narrow interface approach you are establishing here.

EdDev replied:
I can try to narrow this down and see how it looks.
But can we do this as a follow up? I prefer to take this in to have a baseline for further work in this area.

Do we have an example of it used in a narrow manner like this in another place?

xpivarc replied:
I am ok to take this in follow up. Example can be found in `pkg/virt-handler/node-labeller/node_labeller.go`, I think we generally suffer from too broad interfaces :)

EdDev replied:
ack
I'll make sure this will be followed up with.

@orelmisan , would you be able to help improve this as a follow up PR?

orelmisan replied:
Sure.

xpivarc commented on pkg/virt-controller/watch/vm.go:
```diff
@@ -3141,31 +3145,17 @@ func (c *VMController) sync(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachin
-
-		if err := c.vmiInterfacesPatch(&vmiCopy.Spec, vmi); err != nil {
-			return vm, &syncErrorImpl{fmt.Errorf("Error encountered when trying to patch vmi: %v", err), FailedUpdateErrorReason}, nil
+	for _, syncer := range c.Synchronizers {
```
Does the network synchronizer uses a `RestartRequired` condition in any condition?

EdDev replied:
No.
We have not yet considered the consolidation into the rollout concept.

EdDev commented:
[change](https://github.com/kubevirt/kubevirt/compare/756659f3111788862b3dc0d1d2ab45b5ce976d79..553ecc98aa62b1de669963575a71697fcfbebe66): Per the review, removed the pool of synchronizers and replaced it with just a single network one. This is a compromise to allow this change to get in.

EdDev commented:
change: Rebase

EdDev commented:
Some odd CI platform error:

```
 13:32:29: Error response from daemon: cannot listen on the TCP port: listen tcp4 :42607: bind: address already in use
13:32:30: ./cluster-up/up.sh: line 34: pop_var_context: head of shell_variables not a function context
make: *** [Makefile:153: cluster-up] Error 1 
```

Retrying...

/test pull-kubevirt-e2e-k8s-1.30-sig-network

xpivarc commented:
> [change](https://github.com/kubevirt/kubevirt/compare/756659f3111788862b3dc0d1d2ab45b5ce976d79..553ecc98aa62b1de669963575a71697fcfbebe66): Per the review, removed the pool of synchronizers and replaced it with just a single network one. This is a compromise to allow this change to get in.

/approve
Let's sync on the synchronizers to follow up

EdDev commented:
/retest-required
